src/batch/src/executor/join/lookup_join.rs

```diff
@@ -317,263 +316,191 @@ impl<P: 'static + ProbeSideSourceBuilder> Executor for LookupJoinExecutor<P> {
+        Self {
+            join_type,
+            condition,
+            build_child,
+            build_side_data_types,
+            build_side_key_idxs,
+            probe_side_source,
+            probe_side_key_types,
+            probe_side_key_idxs,
+            null_safe,
+            chunk_builder,
+            schema,
+            output_indices,
+            identity,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<K: HashKey> LookupJoinExecutor<K> {
     #[try_stream(boxed, ok = DataChunk, error = RwError)]
     async fn do_execute(mut self: Box<Self>) {
-        let mut build_side_batch_read_stream: BoxedDataChunkListStream = utils::batch_read(
-            self.build_child.take().unwrap().execute(),
-            AT_LEAST_BUILD_SIDE_ROWS,
-        );
+        let lookup_join_build_side_schema = self.build_child.schema().clone();
+        let mut lookup_join_build_side_batch_read_stream: BoxedDataChunkListStream =
+            utils::batch_read(self.build_child.execute(), AT_LEAST_BUILD_SIDE_ROWS);
 
-        while let Some(chunk_list) = build_side_batch_read_stream.next().await {
+        while let Some(chunk_list) = lookup_join_build_side_batch_read_stream.next().await {
             let chunk_list = chunk_list?;
 
             // Group rows with the same key datums together
             let groups = chunk_list
                 .iter()
-                .flat_map(|chunk| chunk.rows())
-                .into_group_map_by(|row| {
-                    self.build_side_key_idxs
-                        .iter()
-                        .map(|&idx| row.value_at(idx).to_owned_datum())
-                        .collect_vec()
-                });
-
-            let mut row_keys = vec![];
-            let mut all_row_refs = vec![];
-
-            groups.into_iter().for_each(|(row_key, row_refs)| {
-                row_keys.push(row_key);
-                all_row_refs.push(row_refs);
-            });
-
-            assert_eq!(row_keys.len(), all_row_refs.len());
+                .flat_map(|chunk| {
+                    chunk.rows().map(|row| {
+                        self.build_side_key_idxs
+                            .iter()
+                            .map(|&idx| row.value_at(idx).to_owned_datum())
+                            .collect_vec()
+                    })
+                })
+                .sorted()
+                .dedup()
+                .collect_vec();
 
             self.probe_side_source.reset();
-            for row_key in &row_keys {
+            for row_key in &groups {
                 self.probe_side_source.add_scan_range(row_key)?;
             }
+            let lookup_join_probe_child = self.probe_side_source.build_source().await?;
+
+            // NOTICE!!!
+            // Lookup join build side will become the probe side of hash join,
+            // while its probe side will become the build side of hash join.
+            let hash_join_probe_child = Box::new(BufferChunkExecutor::new(
```

Comment:
This is quite confusing, maybe we should unfiy the naming 😢

Reply:
Leave it for later PR to do the rename stuff.